src/core/Ekubo.sol

```diff
@@ -0,0 +1,486 @@
+            bytes memory encodedPayedAmount =
+                _setOperatorAndCall(msg.sender, data, uint32(IEkuboCallbacks.payCallback.selector), payCallback);
+            assembly ("memory-safe") {
+                // We can skip all the checks performed by `abi.decode` because we know that this is the
+                // verbatim result from `payCallback` and that `payCallback` encoded the payment
+                // correctly.
+                payment := mload(add(0x60, encodedPayedAmount))
+            }
+        }
+    }
+
+    // the mandatory fields are
+    // 2 - sell bps
+    // 1 - pool key tokens case
+    // 32 - config (20 extension, 8 fee, 4 tickSpacing)
+    // 32 - skipAhead
+    uint256 private constant _HOP_DATA_LENGTH = 67;
+
+    uint256 private constant _ADDRESS_MASK = 0x00ffffffffffffffffffffffffffffffffffffffff;
+
+    function locked(bytes calldata data) private returns (bytes memory) {
+        address recipient;
+        uint256 minBuyAmount;
+        uint256 hashMul;
+        uint256 hashMod;
+        bool feeOnTransfer;
+        address payer;
+        (data, recipient, minBuyAmount, hashMul, hashMod, feeOnTransfer, payer) = Decoder.decodeHeader(data);
+
+        // Set up `state` and `notes`. The other values are ancillary and might be used when we need
+        // to settle global sell token debt at the end of swapping.
+        (
+            bytes calldata newData,
+            StateLib.State memory state,
+            NotesLib.Note[] memory notes,
+            ISignatureTransfer.PermitTransferFrom calldata permit,
+            bool isForwarded,
+            bytes calldata sig
+        ) = Decoder.initialize(data, hashMul, hashMod, payer);
+        if (state.sell.amount >= (1 << 127)) {
+            Panic.panic(Panic.ARITHMETIC_OVERFLOW);
+        }
+        if (payer != address(this)) {
+            state.globalSell.amount = _permitToSellAmountCalldata(permit);
+        }
+        if (feeOnTransfer) {
+            state.globalSell.amount =
+                _ekuboPay(state.globalSell.token, payer, state.globalSell.amount, permit, isForwarded, sig);
+        }
+        if (state.globalSell.amount == 0) {
+            revert ZeroSellAmount(state.globalSell.token);
+        }
+        state.globalSellAmount = state.globalSell.amount;
+        data = newData;
+
+        PoolKey memory poolKey;
+
+        while (data.length >= _HOP_DATA_LENGTH) {
+            uint16 bps;
+            assembly ("memory-safe") {
+                bps := shr(0xf0, calldataload(data.offset))
+
+                data.offset := add(0x02, data.offset)
+                data.length := sub(data.length, 0x02)
+                // we don't check for array out-of-bounds here; we will check it later in `Decoder.overflowCheck`
+            }
+
+            data = Decoder.updateState(state, notes, data);
+            int256 amountSpecified;
+            unchecked {
+                amountSpecified = int256((state.sell.amount * bps).unsafeDiv(BASIS));
```

Comment:
It might be good to briefly note why it's not possible for this operation to overflow